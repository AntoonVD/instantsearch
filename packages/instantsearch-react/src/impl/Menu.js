import React, {PropTypes, Component} from 'react';
import themeable from 'react-themeable';

import {itemsPropType, selectedItemsPropType} from '../propTypes';
import {getTranslation} from '../utils';

import LinkItem from './LinkItem';

const defaultTranslations = {
  showMore: extended => extended ? 'Show less' : 'Show more',
  count: count => count.toString(),
};

const defaultTheme = {
  root: 'Menu',
  extended: 'Menu--extended',
  list: 'Menu__list',
  item: 'Menu__item',
  itemSelected: 'Menu__item--selected',
  itemLink: 'Menu__item__link',
  itemValue: 'Menu__item__value',
  itemCount: 'Menu__item__count',
  showMore: 'Menu__showMore',
};

class Menu extends Component {
  static propTypes = {
    refine: PropTypes.func.isRequired,
    createURL: PropTypes.func.isRequired,
    items: itemsPropType,
    selectedItems: selectedItemsPropType,
    translations: PropTypes.object,
    theme: PropTypes.object,
    showMore: PropTypes.bool,
    limitMin: PropTypes.number,
    limitMax: PropTypes.number,
    limit: PropTypes.number.isRequired,
    show: PropTypes.func.isRequired,
  };

  static defaultProps = {
    translations: defaultTranslations,
    theme: defaultTheme,
    showMore: false,
    limitMin: 10,
    limitMax: 20,
  };

  onShowMoreClick = () => {
    this.props.show(
      this.props.limit === this.props.limitMax ?
        this.props.limitMin :
        this.props.limitMax
    );
  };

  onItemClick = item => {
    this.props.refine(item.value);
  };

  render() {
    const {
      translations,
      theme,
      items,
      selectedItems,
      showMore,
      createURL,
      limit,
      limitMax,
    } = this.props;
    if (!items) {
      return null;
    }

    const th = themeable(theme);

    return (
      <div {...th('root', 'root')}>
        <ul {...th('list', 'list')}>
          {items
            .slice(0, limit)
            .map(item =>
              <li
                {...th(
                  item.value,
                  'item',
                  selectedItems.indexOf(item.value) !== -1 && 'itemSelected'
                )}
              >
                <LinkItem
                  {...th('itemLink', 'itemLink')}
                  onClick={this.onItemClick}
                  item={item}
                  href={createURL(item.value)}
                >
                  <span {...th('itemLabel', 'itemLabel')}>
                    {item.value}
                  </span>
                  {' '}
                  <span {...th('itemCount', 'itemCount')}>
                    {getTranslation(
                      'count',
                      defaultTranslations,
                      translations,
                      item.count
                    )}
                  </span>
                </LinkItem>
              </li>
            )
          }
        </ul>
        {showMore &&
          <button
            {...th('showMore', 'showMore')}
            onClick={this.onShowMoreClick}
          >
            {getTranslation(
              'showMore',
              defaultTranslations,
              translations,
              limit === limitMax
            )}
          </button>
        }
      </div>
    );
  }
}

export default Menu;
