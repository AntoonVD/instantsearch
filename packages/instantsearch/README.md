# instantsearch

Go to the [instantsearch website](https://community.algolia.com/instantsearch.js/v2/vanilla) for more information.
