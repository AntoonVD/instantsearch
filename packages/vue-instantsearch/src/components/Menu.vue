<template>
  <div :class="bem()" v-if="show">
    <slot name="header"></slot>

    <div
      v-for="item in state.items"
      :key="item.label"
      :class="item.isRefined ? bem('item', 'active') : bem('item')"
    >
      <a
        href="#"
        :class="bem('link')"
        @click.prevent="state.refine(item.value)"
      >
        {{item.isRefined ? "x" : ""}} {{item.label}}
        <span :class="bem('count')">{{item.count}}</span>
      </a>
    </div>

    <slot name="footer"></slot>
  </div>
</template>

<script>
import algoliaComponent from '../component';
import { connectMenu } from 'instantsearch.js/es/connectors';

export default {
  mixins: [algoliaComponent],
  props: {
    attribute: {
      type: String,
      required: true,
    },
    limit: {
      type: Number,
      default: 10,
    },
    showMoreLimit: {
      type: Number,
    },
    sortBy: {
      default() {
        return ['isRefined:desc', 'count:desc', 'name:asc'];
      },
    },
  },
  computed: {
    show() {
      return this.state.items.length > 0;
    },
    widgetParams() {
      return {
        attributeName: this.attribute,
        limit: this.limit,
        showMoreLimit: this.showMoreLimit,
        sortBy: this.sortBy,
      };
    },
  },
  data() {
    return {
      blockClassName: 'ais-menu',
    };
  },
  beforeCreate() {
    this.connector = connectMenu;
  },
};</script>
