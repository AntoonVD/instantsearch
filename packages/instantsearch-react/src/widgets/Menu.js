import createMenu from '../hoc/createMenu';
import Menu from '../impl/Menu';
export default createMenu(Menu);
