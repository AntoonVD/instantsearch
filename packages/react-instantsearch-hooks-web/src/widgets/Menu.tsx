import React from 'react';
import { useMenu } from 'react-instantsearch-hooks';

import { Menu as MenuUiComponent } from '../ui/Menu';

import type { MenuProps as MenuUiComponentProps } from '../ui/Menu';
import type { UseMenuProps } from 'react-instantsearch-hooks';

export type MenuProps = Omit<
  MenuUiComponentProps,
  | 'items'
  | 'onRefine'
  | 'createURL'
  | 'noResults'
  | 'canToggleShowMore'
  | 'onToggleShowMore'
  | 'isShowingMore'
> &
  UseMenuProps;

export function Menu({
  attribute,
  limit,
  showMore,
  showMoreLimit,
  sortBy,
  transformItems,
  ...props
}: MenuProps) {
  const {
    canToggleShowMore,
    isShowingMore,
    items,
    refine,
    createURL,
    toggleShowMore,
  } = useMenu(
    {
      attribute,
      limit,
      showMore,
      showMoreLimit,
      sortBy,
      transformItems,
    },
    {
      $$widgetType: 'ais.menu',
    }
  );

  return (
    <MenuUiComponent
      {...props}
      items={items}
      createURL={createURL}
      onRefine={(item) => {
        refine(item.value);
      }}
      showMore={showMore}
      canToggleShowMore={canToggleShowMore}
      onToggleShowMore={toggleShowMore}
      isShowingMore={isShowingMore}
    />
  );
}
