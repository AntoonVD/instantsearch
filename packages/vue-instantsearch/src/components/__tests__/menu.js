import Vue from 'vue';
import instantsearch from 'instantsearch.js/es/';

import Menu from '../Menu.vue';
import { FACET_TREE } from '../../store';

const instance = instantsearch({
  appId: '...',
  apiKey: '...',
  indexName: '...',
});

const getFacetValues = jest.fn(attributeName => ({
  name: attributeName,
  count: null,
  isRefined: true,
  path: null,
  data: [
    {
      count: 319,
      data: null,
      isRefined: false,
      name: 'Bathroom',
      path: 'Bathroom',
    },
    {
      count: 100,
      data: null,
      isRefined: true,
      name: 'Clothes',
      path: 'Clothes',
    },
    {
      count: 202,
      data: null,
      isRefined: false,
      name: 'Shoes',
      path: 'Shoes',
    },
    {
      count: 398,
      data: null,
      isRefined: false,
      name: 'Kitchen',
      path: 'Kitchen',
    },
  ],
}));

const searchStore = {
  getFacetValues,
  addFacet: () => {},
  setMaxValuesPerFacet: () => {},
  stop: () => {},
  start: () => {},
  refresh: () => {},
};

describe.skip('Menu', () => {
  it('should render correctly', () => {
    const Component = Vue.extend(Menu);
    const vm = new Component({
      propsData: {
        attribute: 'category',
        searchStore,
        instance,
      },
    });
    vm.$mount();
    expect(vm.$el.outerHTML).toMatchSnapshot();
  });

  it('should render nothing if no menu entries are available', () => {
    const Component = Vue.extend(Menu);

    const store = Object.assign({}, searchStore, {
      getFacetValues: () => ({ data: null }),
    });
    const vm = new Component({
      propsData: {
        attribute: 'category',
        searchStore: store,
        instance,
      },
    });
    vm.$mount();
    expect(vm.$el.outerHTML).toMatchSnapshot();
  });

  it('should add a facet to the store when created', () => {
    const Component = Vue.extend(Menu);
    const addFacetMock = jest.fn();

    // eslint-disable-next-line no-new
    new Component({
      propsData: {
        attribute: 'category',
        searchStore: Object.assign({}, searchStore, {
          addFacet: addFacetMock,
          setMaxValuesPerFacet: () => {},
        }),
        instance,
      },
    });

    expect(addFacetMock).toBeCalledWith(
      {
        name: 'category',
        attributes: ['category'],
      },
      FACET_TREE
    );
  });

  it('should remove the facet from the store when destroyed', () => {
    const Component = Vue.extend(Menu);
    const removeFacetMock = jest.fn();

    const vm = new Component({
      propsData: {
        attribute: 'category',
        searchStore: Object.assign({}, searchStore, {
          removeFacet: removeFacetMock,
        }),
        instance,
      },
    });

    vm.$mount();
    vm.$destroy();

    expect(removeFacetMock).toBeCalledWith('category');
  });

  it('should set `maxValuesPerFacet` when mounted', () => {
    const Component = Vue.extend(Menu);
    const store = Object.assign({}, searchStore);

    // eslint-disable-next-line no-new
    new Component({
      propsData: {
        limit: 10,
        attribute: 'category',
        searchStore: store,
        instance,
      },
    });

    expect(store.maxValuesPerFacet).toBe(10);
  });
});
