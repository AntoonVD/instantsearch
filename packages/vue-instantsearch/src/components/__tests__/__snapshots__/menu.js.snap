// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render correctly 1`] = `

<div class="ais-menu">
  <div class="ais-menu__item">
    <a class="ais-menu__link"
       href="#"
    >
      Bathroom
      <span class="ais-menu__count">
        319
      </span>
    </a>
  </div>
  <div class="ais-menu__item">
    <a class="ais-menu__link"
       href="#"
    >
      Clothes
      <span class="ais-menu__count">
        100
      </span>
    </a>
  </div>
  <div class="ais-menu__item">
    <a class="ais-menu__link"
       href="#"
    >
      Shoes
      <span class="ais-menu__count">
        202
      </span>
    </a>
  </div>
  <div class="ais-menu__item">
    <a class="ais-menu__link"
       href="#"
    >
      Kitchen
      <span class="ais-menu__count">
        398
      </span>
    </a>
  </div>
</div>

`;
